testing/runner.ts

```diff
@@ -95,74 +93,110 @@ export async function getMatchingUrls(
 
   return matchingLocalUrls.concat(matchingRemoteUrls);
 }
+
+export interface RunTestModulesOptions extends RunOptions {
+  exclude?: string[];
+}
+
 /**
- * This function runs matching test files in `root` directory.
+ * Import the specified test modules and run their tests as a suite.
+ *
+ * Test modules are specified as an array of strings and can include local files
+ * or URLs.
  *
- * File matching and excluding supports glob syntax, ie. if encountered arg is
- * a glob it will be expanded using `glob` method from `fs` module.
+ * File matching and excluding support glob syntax - arguments recognized as
+ * globs will be expanded using `glob()` from the `fs` module.
  *
- * Note that your shell may expand globs for you:
- *    $ deno -A ./runner.ts **\/*_test.ts **\/test.ts
+ * Example:
  *
- * Expanding using `fs.glob`:
- *    $ deno -A ./runner.ts \*\*\/\*_test.ts \*\*\/test.ts
+ *       runTestModules(["**\/*_test.ts", "**\/test.ts"]);
  *
- *  `**\/*_test.ts` and `**\/test.ts"` are arguments that will be parsed and
- *  expanded as: [glob("**\/*_test.ts"), glob("**\/test.ts")]
+ * Any matched directory `<dir>` will expand to:
+ *   - `<dir>/**\/*_test.ts`
+ *   - `<dir>/**\/*_test.js`
+ *   - `<dir>/**\/test.ts`
+ *   - `<dir>/**\/test.js`
+ *
+ * So the above example is captured naturally by:
+ *
+ *       runTestModules(["."]);
  */
-// TODO: change return type to `Promise<void>` once, `runTests` is updated
-// to return boolean instead of exiting
-export async function main(root: string = cwd()): Promise<void> {
+// TODO: Change return type to `Promise<void>` once, `runTests` is updated
+// to return boolean instead of exiting.
+export async function runTestModules(
+  include: string[],
```

Comment:
-1 for having `include` as first arg and the rest in options. I think it makes sense to put them all inside `options`:
```
runTestModules({include: ["."], exclude: ["dist/"]})
```

Reply:
Hmm, I want to separate out the optional arguments into `RunTestModuleOptions`. This function doesn't make sense without includes, whereas the excludes wouldn't be used for most well structured source trees. Also it's analogous to the CLI.

Reply:
I'm not sure... If I run `deno test` it should work like now: walk `root` and find all files matching default globs. So you should be able to run just `runTestModules()` without any arguments.

Reply:
Ah okay, you want to "." default to apply to the exported API and not just the CLI.

I think an optional array argument that default to a singleton array isn't a very intuitive design. You have `runTestModules()` and `runTestModules({ include: [] })` as two situations where the default could (possibly?) be applied... it just complicates the idea of a default. The API ends up being kind of hard to reason about and the implementation would be ugly as well.

Reply:
> I think an optional array argument that default to a singleton array isn't a very intuitive design. You have `runTestModules()` and `runTestModules({ include: [] })` as two situations where the default could (possibly?) be applied... it just complicates the idea of a default. The API ends up being kind of hard to reason about and the implementation would be ugly as well.

I'm not sure I follow...

If you defined `runTestModules` as such:
```
function runTestModules(options: {
  include?: string[],
  exclude?: string[]
} = {}) {
...
options = {...DEFAULT_OPTIONS, ...options};
}
```
Then the API shouldn't be that hard.

Reply:
Ah! Now I see your point.

Then maybe we should split options for globs and directories to find?

If we keep options like now, then we'd have something alike:
```ts
// I'd expect to find all files matching default globs inside `src/` directory relative to current working directory.
runTestModules({ root: "src/" });

// I'd expect to find all files matching provided globs relative to current working directory.
runTestModules({ include: ["**/*.test.ts", "**/*.test.js"] });

// I'd expect to run only that single file without any walking of directory tree.
runTestModules({ include: ["https://deno.land/std/http/file_server_test.ts"] });

// I'd expect it would find files matching default globs 
runTestModules({ include: [] })
```

But then `include` (and `exclude` as well) would sometimes be used to specify globs, sometimes directory path (still applying globs) and sometimes remote files.

🤔

Reply:
It's high time to check how other runners (Ava, Mocha, Jest) handle that. The problem is AFAIK neither of them can run remote files.

Reply:
Yes, glob expansion is done post-partition. They won't matter for this change, hopefully. I think URLs are only supported here for consistency with Deno's definition of a module.

Reply:
@bartlomieju 😅 I've come around to your original suggestion.

My thought is that this function is not really a good logical unit and will pretty much only be used as an exact stand-in for the CLI version. It should be as analogous to the CLI as possible since that's the only value it has. Ignore my rambling above :p